src/LogOtter.CosmosDb.EventStore/EventStore/EventStore.cs

```diff
@@ -107,9 +129,36 @@ public async Task<IReadOnlyCollection<StorageEvent>> ReadStreamForwards(string s
         return events.AsReadOnly();
     }
 
-    public StorageEvent FromCosmosStorageEvent(CosmosDbStorageEvent cosmosDbStorageEvent)
+    public async Task<int> ReadStreamEventCount(string streamId)
+    {
+        var requestOptions = new QueryRequestOptions { PartitionKey = new PartitionKey(streamId) };
+        return await _container
+            .GetItemLinqQueryable<CosmosDbStorageEvent>(requestOptions: requestOptions)
+            .Where(e => e.StreamId == streamId)
+            .CountAsync();
+    }
+
+    public async Task<StorageEvent<TBaseEvent>> ReadEventFromStream(string streamId, Guid eventId, CancellationToken cancellationToken = default)
+    {
+        var requestOptions = new QueryRequestOptions { PartitionKey = new PartitionKey(streamId) };
+        var query = _container
+            .GetItemLinqQueryable<CosmosDbStorageEvent>(requestOptions: requestOptions)
+            .Where(e => e.StreamId == streamId && e.EventId == eventId);
+
+        var feedIterator = _feedIteratorFactory.GetFeedIterator(query);
+
+        if (!feedIterator.HasMoreResults)
+        {
+            throw new CosmosException($"Event {eventId} not found for stream '{streamId}'", HttpStatusCode.Conflict, 0, null, 0);
+        }
+
+        var response = await feedIterator.ReadNextAsync(cancellationToken);
```

Comment:
`ReadNextAsync` can return zero results even when `feedIterator.HasMoreResults` is `true` - so you will need to keep looping until you get an item.

Reply:
will do